pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -759,16 +765,141 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
+	conn *dns.Conn
+
+	// this mutex protects writes on 'conn' and all access to 'reqs'
+	lock.RWMutex
+	reqs map[uint16]chan Response // outstanding requests
+
+	// 'reader' mutex is used to serialize reads on 'conn'. It is always taken and released
+	// while holding the lock but the lock can be released and re-acquired while holding
+	// 'reader' mutex.
+	reader lock.Mutex
+}
+
+// ExchangeAsync writes the request to the Client's connection and co-operatively
+// reads responses from the connection and distributes them to the requestors.
+// At most one caller is reading from Client's connection at any time.
+func (c *Client) ExchangeAsync(m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	// Lock allows only one request to be written at a time, but that can happen
+	// concurrently with reading.
+	c.Lock()
+	defer c.Unlock()
+	if _, exists := c.reqs[m.Id]; exists {
+		return nil, 0, fmt.Errorf("Duplicate request: %d", m.Id)
+	}
+
+	// Send while holding the client lock, as dns.Client is not made to be usable from
+	// concurrent goroutines.
+	start := time.Now()
+	err = c.SendContext(context.Background(), m, c.conn, start)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Create channel for the response with buffer of one, so that write to it
+	// does not block if we happen to do it ourselves.
+	ch := make(chan Response, 1)
+	c.reqs[m.Id] = ch
+
+	// Wait for the response
+	var resp Response
+	for {
+		// Try taking the reader lock
+		if c.reader.TryLock() {
+			// We are responsible for reading responses for all users
+			// of this client until we get our own response or an error occurs.
+			var err error
+			for err == nil {
+				// Release the client lock for the duration of the blocking read
+				// operation to allow concurrent writes to the underlying
+				// connection.
+				var r *dns.Msg
+				c.Unlock()
+				r, err = c.conn.ReadMsg()
+				c.Lock()
+				if err != nil {
+					break
+				}
+				// Locate the request for this response, skipping if not found
+				ch, exists := c.reqs[r.Id]
+				if !exists {
+					continue
+				}
+				// Pass the response to the waiting requester
+				delete(c.reqs, r.Id)
+				ch <- Response{Msg: r}
+				if r.Id == m.Id {
+					// Got our response, quit reading and tell others that
+					// its their turn to read.
+					err = errNoReader
+				}
+			}
+			// Releasing the reader lock before sending errors on waiter's channels
+			// so that when they get them, one of them can take the reader lock.
+			c.reader.Unlock()
+			for id, ch := range c.reqs {
+				// Another reader will pick up if any errNoReader errors are sent.
+				// Only delete the pending request in other error cases.
+				if !errors.Is(err, errNoReader) {
+					delete(c.reqs, id)
+				}
+				ch <- Response{err: err}
+			}
+		}
+		// Get the response of error from the current reader.
+		// Unlock for the blocking duration to allow concurrent writes
+		// on the client's connection.
+		c.Unlock()
+		resp = <-ch
```

Comment:
Why doesn't this block a reader who lost the `TryLock` forever if there is no response for the request ID it is waiting for? Would that leak a goroutine?

Reply:
Each reader takes turns on reading for as long as it gets its response, or connection of the `dns.Client` fails. `dns` package sets deadlines on the connection on every read, so those will fail the connection if there is no response. In this regard this "co-operative" reading behavior is no different from a single blocking call to `c.conn.ReadMsg()`.

from Go docs:
```
    // SetReadDeadline sets the deadline for future Read calls.
    // A zero value for t means Read will not time out.
    SetReadDeadline(t time.Time) error

    // SetWriteDeadline sets the deadline for future Write calls.
    // Even if write times out, it may return n > 0, indicating that
    // some of the data was successfully written.
    // A zero value for t means Write will not time out.
    SetWriteDeadline(t time.Time) error
```
From this I read that the deadline will affect all future (read) calls, not just the next one.

Setting of the deadline is here: https://github.com/cilium/dns/blob/master/client.go#L249-L265

It gets configured here: https://github.com/cilium/cilium/blob/362b80ef8548db4ba4185f4df8113f0c217a9ac8/pkg/fqdn/dnsproxy/proxy.go#L1074

Reply:
Added this comment:
```
				// This ReadMsg() will eventually fail due to the read deadline set
				// by 'dns.Client' on the underlying connection when sending the
				// (last) request.
				r, err = c.conn.ReadMsg()
```

Reply:
Ah- I was asking about the `ch` read, but while trying to clarify my question I found my mistake- I missed that on getting their response the reading goroutine will send an error to the channels.